src/contract/contract.ts

```diff
@@ -26,165 +26,310 @@ import { node } from '../api'
+  artifact: T
 
-export abstract class Common {
-  readonly sourceCodeSha256: string
-  readonly functions: FunctionSig[]
+  constructor(sourceFile: SourceFile, artifact: T) {
+    this.sourceFile = sourceFile
+    this.artifact = artifact
+  }
+}
+
+class ProjectArtifact {
+  static readonly artifactFileName = '.project.json'
+
+  files: Map<string, string> // file path -> source code hash
+
+  constructor(files: Map<string, string>) {
+    this.files = files
+  }
+
+  async saveToFile(rootPath: string): Promise<void> {
+    const filepath = rootPath + '/' + ProjectArtifact.artifactFileName
+    const content = JSON.stringify(Object.fromEntries(this.files), null, 2)
+    return fsPromises.writeFile(filepath, content)
+  }
 
-  static readonly importRegex = new RegExp('^import "([^"/]+/(([^"]+)/)?)?[a-z][a-z_0-9]*.ral"', 'mg')
-  static readonly contractRegex = new RegExp('^(Abstract[ ]+)?Contract [A-Z][a-zA-Z0-9]*', 'mg')
-  static readonly interfaceRegex = new RegExp('^Interface [A-Z][a-zA-Z0-9]* \\{', 'mg')
-  static readonly scriptRegex = new RegExp('^TxScript [A-Z][a-zA-Z0-9]*', 'mg')
-
-  private static _artifactCache: Map<string, Contract | Script> = new Map<string, Contract | Script>()
-  static artifactCacheCapacity = 20
-  protected static _getArtifactFromCache(codeHash: string): Contract | Script | undefined {
-    return this._artifactCache.get(codeHash)
-  }
-  protected static _putArtifactToCache(contract: Contract): void {
-    if (!this._artifactCache.has(contract.codeHash)) {
-      if (this._artifactCache.size >= this.artifactCacheCapacity) {
-        const keyToDelete = this._artifactCache.keys().next().value
-        this._artifactCache.delete(keyToDelete)
+  sourceHasChanged(files: SourceFile[]): boolean {
+    if (files.length != this.files.size) {
+      return true
+    }
+    for (const file of files) {
+      const sourceCodeHash = this.files.get(file.contractPath)
+      if (typeof sourceCodeHash === 'undefined' || sourceCodeHash != file.sourceCodeHash) {
+        return true
       }
-      this._artifactCache.set(contract.codeHash, contract)
     }
+    return false
   }
 
-  constructor(sourceCodeSha256: string, functions: FunctionSig[]) {
-    this.sourceCodeSha256 = sourceCodeSha256
-    this.functions = functions
+  static async from(rootPath: string): Promise<ProjectArtifact | undefined> {
+    const filepath = rootPath + '/' + ProjectArtifact.artifactFileName
+    if (!fs.existsSync(filepath)) {
+      return undefined
+    }
+    const content = await fsPromises.readFile(filepath)
+    const files = new Map(Object.entries<string>(JSON.parse(content.toString())))
+    return new ProjectArtifact(files)
   }
+}
 
-  protected static _artifactsFolder(): string {
-    return './artifacts/'
+export class Project {
+  sourceFiles: SourceFile[]
+  contracts: ArtifactInfo<Contract>[]
+  scripts: ArtifactInfo<Script>[]
+
+  private static instance: Project
+  private static nodeProvider: NodeProvider
```

Comment:
`nodeProvider` can be singleton, but I am not sure about `instance`.
1. we might support subprojects for any source directory, instead of `contracts`
2. one might create multiple projects.

Maybe we could load a project when `fromSource` is called?

Reply:
Another reason I made it a singleton is because the `Contract` depends on the `Project`, and the contract needs to find other contracts by the contract code hash, which is used in `testPublicMethod/testPrivateMethod`.

If we want to support subprojects, I guess we need to add:

1. Add a `Project.fromSource(dirPath)` method used to load the project
2. Add a parameter to the contract `testPublicMethod/testPrivateMethod` methods, so it can get the other contracts by contract code hash.

WDYT?

Reply:
I made these fields non-static in the latest commit, and now we can support creating multiple projects.